### ISSUE TITLE
Database/Type/DateTimeType not respecting DB and PHP timezones.

### ISSUE THREAD

**Issue by jhawkwind** [cakephp/cakephp#6571]
If in your configuration you specify one timezone for the DB settings, and one timezone for the PHP settings, the time value that gets passed from the DB server through the client, then through PHP will simply pass text over, ignoring that all three time zones could be different and need conversion of the actual time value.

Let me provide a sample code first as it is easier to illustrate this way. Currently the code looks like this and just passes the value into a DateTime object instance:

``` php
    /**
     * Convert strings into DateTime instances.
     *
     * @param string $value The value to convert.
     * @param Driver $driver The driver instance to convert with.
     * @return \Cake\I18n\Time|DateTime
     */
    public function toPHP($value, Driver $driver)
    {
        if ($value === null || strpos($value, '0000-00-00') === 0) {
            return null;
        }

        if (strpos($value, '.') !== false) {
            list($value) = explode('.', $value);
        }

        $instance = clone $this->_datetimeInstance;
        return $instance->modify($value);
    }
```

Recommended changes are like this to take into different timezone settings into account:

``` php
    /**
     * Convert strings into DateTime instances.
     *
     * @param string $value The value to convert.
     * @param Driver $driver The driver instance to convert with.
     * @return \Cake\I18n\Time|DateTime
     */
    public function toPHP($value, Driver $driver)
    {
        if ($value === null || strpos($value, '0000-00-00') === 0) {
            return null;
        }

        if (strpos($value, '.') !== false) {
            list($value) = explode('.', $value);
        }

        $instance = new static::$dateTimeClass($value,$this->_DBTimeZone);
        $instance->modify($value);
        $instance->setTimeZone($this->_PHPTimeZone);
        return $instance;
    }
```

The short version is that, this will take into account the timezone settings you may have set in the application configuration for the DB and recognize that the value passed from the DB belong to a certain timezone that is different from PHP.

So let's say the DB client is configured for `America/Los_Angeles`; it will figure out what timezone you have set as default in PHP, let's say `America/New_York`; then it will take the DB value it recieves, advance the time 3 hours, and then present it for use for the PHP code that expects all times to be in **Eastern Time**.

This is especially needed since the DB Server could be set to `UTC`, when the client connection is configured for anything but the server timezone, the client (in this case the MySQL client) will convert from `UTC` to **Pacific Time** before presenting the value to you. And under the current code, it assumes always that it is UTC, which is entirely wrong, since the DB has changed the UTC value to a `America/Los_Angeles` value.

The reverse needs to be true also, if a DATETIME object is passed, it needs to take into account the timezone assigned, and set the value to the DB timezone before extracting the string to pass to the DB. And if a string is passed to be placed into the DB, then the default PHP timezone needs to be taken into account and changed to match the DB time.

Maybe something changed to this:

``` php
/**
     * Convert DateTime instance into strings.
     *
     * @param string|int|\DateTime $value The value to convert.
     * @param Driver $driver The driver instance to convert with.
     * @return string
     */
    public function toDatabase($value, Driver $driver)
    {
        $newValue = null;

        if($value === null) {
            $newValue = null;
        }
        if(is_string($value)) {
            $valueObj = new static::$dateTimeClass($value,$this->_PHPTimeZone);
            $valueObj->setTimezone($this->_DBTimeZone);
            $newValue = $valueObj->format($this->_format);
        }
        elseif (is_int($value)) {
            $value = new static::$dateTimeClass('@' . $value,$this->_PHPTimeZone);
            $valueObj->setTimezone($this->_DBTimeZone);
            $newValue = $valueObj->format($this->_format);
        }
        elseif ($value instanceof static::$dateTimeClass) {
            $value->setTimezone($this->_DBTimeZone);
            $newValue = $valueObj->format($this->_format);
        }


        return $newValue;
    }
```

Though I would like some input as to whether there are other considerations that I am missing regarding this. I am inclined to make a new Database/Type instead called: DateTimeWithTimezoneType. So that if you don't want the timezones to recalculate the value then you have the option to opt out.

**Comment by lorenzo** [comment]
A new `DateTimeWithTimezoneType` sounds like the way to go

**Comment by jhawkwind** [comment]
@lorenzo , so aside from creating the DB type file, how would the new type be incorporated and triggered in?

I understand PostgreSQL has a specific column type for `timestamp w/o timezone` and `timestamp with timezone`, and MySQL will internally always convert everything to UTC, but present it as per the client setting. How and when should it be decided whether `DateTimeType` or `DateTimeWithTimezoneType` be used? Suggestions?

**Comment by lorenzo** [comment]
I would say we let that decision to the user by the means of the `_initializeSchema()` method in the Table classes

**Comment by lorenzo** [comment]
on second thought, I think putting this directly into `DateTimeType` is better. I cannot think of a reason where having a timezone mismatch is something you would want. @jhawkwind would you be able to provide a patch for this?

**Comment by markstory** [comment]
Is this only an issue with TIMESTAMP columns, or also with DATETIME columns? Timestamp columns in MySQL are extra funny regarding timezones.

**Comment by lorenzo** [comment]
Mainly DATETIME (although I will need to check) the problem is that php datetime objects should be converted to the target timezone in the database.

**Comment by markstory** [comment]
That could get tricky, I don't think the database abstraction layers should be converting timezones when generating SQL. If you are inserting string data, you probably want _that_ data to go in. It feels presumptuous of the ORM to know better than the human.

**Comment by lorenzo** [comment]
It is not string data @markstory, it is when saving / fetching a datetime object.

**Comment by markstory** [comment]
From reading the proposed code changes string values would be rezoned before being persisted as the toDatabase() was changed.

**Comment by jhawkwind** [comment]
So, specifically MySQL, and with data coming FROM the DB:
With the columns I am currently working with, they are indeed TIMESTAMP, which MySQL will store in `UTC` internally and, as stated in MySQL documentation, will convert depending on your timezone setting for the datasource in config/app.php, what actually gets sent to you, changes. So if I set in `config/app.php` that timezone is `America/Los_Angeles`, it will subtract 7/8 hours before presenting it to you.

The fact that the original code gives you a datetime object and then declares it `UTC` in its timezone field (because it is unspecified it uses the PHP timezone, which could be different) is already a lie; especially if the server time was set to anything but `UTC`. So at the very least, coming FROM the database, one of the following needs to be true:
1.) it needs to put the timezone for the DB that it was set in `config/app.php`;
2.) simply output a string; or
3.) do not set anything in the timezone field (I.E. blank) in the datetime object.

MySQL has the bad habit of making assumptions about the timezone to be the same for everything you supply within the same column type of a given session for `TIMESTAMP`, or all sessions for `DATETIME`. And in almost all cases, specifying the timezone generally gets ignored.

PostgreSQL seems not to be an issue, except for the same fact that the DB client connection specification needs to be respected, since if `TIMESTAMP WITH TIMEZONE` is used, there will be, again, conversion happening if a time is sent in to the DB **without timezone** but into a **with timezone** column. Or if data is pulled out from the DB from a `TIMESTAMP WITHOUT TIMEZONE` column, you need to be able to define what timezone the DB or client is set at and then apply it into the `datetime` object correctly.

In all cases, the best place where this shows up is when DB server timezone setting is not the same as the PHP timezone setting. And worst, is when the DB client interface is set to a third option.

And as a logical extension, with data coming out as timezone defined with conversion, I would at least expect it to also respect the timezone going in to the DB and with operational consistency, if I specified a timezone different from the DB connection. So depending on what we do with DB data output (toPHP), it may impact with what we do with DB data input (toDatabase).

I only have access, at this time, to MySQL and PostgreSQL, so if anyone has other DB flavors to test this with, it would be nice.

**Comment by jhli** [comment]
I'd love to see annoption to disable automatic time zone management all together for a field/an entity/the whole app (thus then also forms) to be sure no magic at all is happening!

**Comment by dereuromark** [comment]
Seems to be related to https://github.com/cakephp/cakephp/issues/10877

**Comment by dereuromark** [comment]
@ADmad Is this something we could solve now in 4.x as well?

**Comment by ADmad** [comment]
If there's clarity about what needs to be done then making required changes shouldn't be an issue. But I fear that the addition of timezone juggling could result in undesirable/unexpected result for someone.

One way would be to introduce a new type class in 4.0 and based on people's feedback make it default in 5.0.

**Comment by ADmad** [comment]
Reading the issue again it's about handling difference in timezone of PHP and database. I misread it earlier and thought it was about dealing with users's timezone.

Part of this issue is already solve by #11780 which adds `setTimezone()` method. The timezone thus set is used to set timezone for the datetime object before generating datetime string for saving to database.

The complimentary part of using the same timezone when generating datetime object from string inside `toPHP()` is what's missing. We can fix that in 4.0.

**Comment by ADmad** [comment]
Closing as related PR is open.
